### PR TITLE
Add wire disjoint check for n-ctrl gate support in LQ

### PR DESCRIPTION
**Pull request by multiphaseCFD**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

[SC-76347]

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp [hunk 1]
```diff
@@ -420,6 +425,11 @@ class StateVectorLQubit : public StateVectorBase<PrecisionT, Derived> {
                         const std::vector<std::size_t> &wires, bool inverse,
                         const std::vector<PrecisionT> &params,
                         const std::vector<ComplexT, Alloc> &matrix) {
+        // Add disjoint check.
```
Comment (josephleekl) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:428]: remove comment?

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp [hunk 2]
```diff
@@ -369,6 +369,11 @@ class StateVectorLQubit : public StateVectorBase<PrecisionT, Derived> {
                         const std::vector<std::size_t> &wires,
                         bool inverse = false,
                         const std::vector<PrecisionT> &params = {}) {
+        // Add disjoint check.
```
Comment (josephleekl) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:372]: do we need this comment?

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp [hunk 3]
```diff
@@ -569,6 +579,10 @@ class StateVectorLQubit : public StateVectorBase<PrecisionT, Derived> {
                           const std::vector<bool> &controlled_values,
                           const std::vector<std::size_t> &wires,
                           bool inverse = false) {
+        // Add disjoint check.
```
Comment (josephleekl) [inline pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:582]: same

**Comment by josephleekl** [comment]
Thanks @multiphaseCFD! This looks good to me, just commented on the added comments

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp (hunk 1 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp]: Done

**Review by josephleekl — APPROVED** [review]
LGTM, this is a good check, thanks @multiphaseCFD!

File: pennylane_lightning/core/src/utils/Util.hpp [hunk 4]
```diff
@@ -574,4 +574,22 @@ std::vector<T1> cast_vector(const std::vector<T0> &vec) {
...
+ */
+template <typename T = std::size_t>
+bool areVecsDisjoint(const std::vector<T> &v1, const std::vector<T> &v2) {
+    std::set<T> s0(v1.begin(), v1.end());
+    for (const auto &element : v2) {
+        if (s0.find(element) != s0.end()) {
+            return false;
+        }
+    }
+    return true;
```
Comment (LuisAlfredoNu) [inline pennylane_lightning/core/src/utils/Util.hpp:593]:
```suggestion
    std::set<int> s0(v1.begin(), v1.end());
    return std::none_of(v2.begin(), v2.end(), [&](int element) {
        return s0.find(element) != s0.end();
    });
```

**Review by LuisAlfredoNu — COMMENTED** [review]
Thank you @multiphaseCFD 
A few comments 
Could you explain a bit why this change, please?

File: pennylane_lightning/core/src/utils/Util.hpp (hunk 4 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/utils/Util.hpp:593]: Could you please elaborate more about the suggestions? Would it lead performance regession?

File: pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp [hunk 5]
```diff
@@ -569,6 +577,9 @@ class StateVectorLQubit : public StateVectorBase<PrecisionT, Derived> {
                           const std::vector<bool> &controlled_values,
                           const std::vector<std::size_t> &wires,
                           bool inverse = false) {
+        PL_ABORT_IF_NOT(
+            areVecsDisjoint<std::size_t>(controlled_wires, wires),
+            "`controlled_wires` and `target wires` must be disjoint.");
```
[the remark this answers is not in the document]
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/simulators/lightning_qubit/StateVectorLQubit.hpp:582]: Let's keep it as it is to give the correct error info.

**Comment by multiphaseCFD** [comment]
> Thank you @multiphaseCFD A few comments Could you explain a bit why this change, please?

Good question @LuisAlfredoNu . Current LQ implement has no safety check for the controlled gates, which users can pass wrong wire info without any warning. For example,  the control_wires and target wires are the same, which is not physically correct, in a few tests in the unit tested I revised.

File: pennylane_lightning/core/src/utils/Util.hpp (hunk 4 above)
Comment (LuisAlfredoNu) [reply pennylane_lightning/core/src/utils/Util.hpp:593]: Do not worry. If we have a regression in the performance it is better to keep the current implementation. I thought that using std algorithm would achieve some performance. Thank you for corroborating the suggestion.  :rocket:

**Review by LuisAlfredoNu — APPROVED** [review]
Thank you @multiphaseCFD 
Nice work and good catch on solving this potential bug 
:rocket: :bug:

**Pull request merged**